src/mongoc/mongoc-cluster.c

```diff
@@ -1794,51 +1858,20 @@ _mongoc_cluster_min_of_max_msg_size_nodes (void *item,
  *--------------------------------------------------------------------------
  */
 int32_t
-mongoc_cluster_get_max_bson_obj_size (mongoc_cluster_t *cluster,
-                                      uint32_t         *server_id /* IN */)
+mongoc_cluster_get_max_bson_obj_size (mongoc_cluster_t *cluster)
 {
-   mongoc_server_description_t *sd;
-   mongoc_cluster_node_t *node;
-   mongoc_topology_t *topology;
    int32_t max_bson_obj_size = -1;
-   bool single_threaded;
-
-   topology = cluster->client->topology;
-   mongoc_mutex_lock (&topology->mutex);
-   single_threaded = topology->single_threaded;
-   mongoc_mutex_unlock (&topology->mutex);
-
-   if (server_id) {
-      if (!single_threaded) {
-         /* if we can, use information stored in cluster node */
-         node = mongoc_set_get (cluster->nodes, *server_id);
-         if (node) {
-            max_bson_obj_size = node->max_bson_obj_size;
-         }
-      } else {
-         /* otherwise, check the server description */
-         mongoc_mutex_lock (&topology->mutex);
-         sd = mongoc_set_get (topology->description.servers, *server_id);
-         if (sd) {
-            max_bson_obj_size = sd->max_bson_obj_size;
-         }
-         mongoc_mutex_unlock (&topology->mutex);
-      }
-   }
 
-   if (max_bson_obj_size == -1) {
-      max_bson_obj_size = MONGOC_CLUSTER_DEFAULT_BSON_OBJ_SIZE;
+   max_bson_obj_size = MONGOC_DEFAULT_BSON_OBJ_SIZE;
 
-      /* if no server was found or given, get min of all values */
-      if (!single_threaded) {
-         mongoc_set_for_each (cluster->nodes,
-                              _mongoc_cluster_min_of_max_obj_size_nodes,
-                              &max_bson_obj_size);
-      } else {
-         mongoc_set_for_each (topology->description.servers,
-                              _mongoc_cluster_min_of_max_obj_size_sds,
-                              &max_bson_obj_size);
-      }
+   if (!cluster->client->topology->single_threaded) {
+      mongoc_set_for_each (cluster->nodes,
+                           _mongoc_cluster_min_of_max_obj_size_nodes,
+                           &max_bson_obj_size);
+   } else {
+      mongoc_set_for_each (cluster->client->topology->description.servers,
```

Comment:
I know we were doing it before, but I don't think this is thread safe.  Other uses of for_each are probably error prone as well